client/pubsub.go

```diff
@@ -3,61 +3,80 @@ package client
 	}
 
-	envelop := &pb.PublishEventRequest{
+	request := &pb.PublishEventRequest{
 		PubsubName: pubsubName,
 		Topic:      topicName,
-		Data:       data,
+	}
+	for _, opt := range opts {
+		opt(request)
+	}
+
+	if data != nil {
+		switch d := data.(type) {
+		case []byte:
+			request.Data = d
+		case string:
+			request.Data = []byte(d)
+		default:
+			var err error
+			request.DataContentType = "application/json"
+			request.Data, err = json.Marshal(d)
+			if err != nil {
+				return errors.WithMessage(err, "error serializing input struct")
+			}
+		}
 	}
 
-	_, err := c.protoClient.PublishEvent(c.withAuthToken(ctx), envelop)
+	_, err := c.protoClient.PublishEvent(c.withAuthToken(ctx), request)
 	if err != nil {
 		return errors.Wrapf(err, "error publishing event unto %s topic", topicName)
 	}
 
 	return nil
 }
 
-// PublishEventfromCustomContent serializes an struct and publishes its contents as data (JSON) onto topic in specific pubsub component.
-func (c *GRPCClient) PublishEventfromCustomContent(ctx context.Context, pubsubName, topicName string, data interface{}) error {
-	if pubsubName == "" {
-		return errors.New("pubsubName name required")
-	}
-	if topicName == "" {
-		return errors.New("topic name required")
-	}
-
-	bytes, err := json.Marshal(data)
-
-	if err != nil {
-		return errors.WithMessage(err, "error serializing input struct")
+// PublishEventWithContentType can be passed as option to PublishEvent to set an explicit Content-Type.
+func PublishEventWithContentType(contentType string) PublishEventOption {
+	return func(e *pb.PublishEventRequest) {
+		e.DataContentType = contentType
 	}
+}
 
-	envelop := &pb.PublishEventRequest{
-		PubsubName:      pubsubName,
-		Topic:           topicName,
-		Data:            bytes,
-		DataContentType: "application/json",
+// PublishEventWithMetadata can be passed as option to PublishEvent to set metadata.
+func PublishEventWithMetadata(metadata map[string]string) PublishEventOption {
+	return func(e *pb.PublishEventRequest) {
+		e.Metadata = metadata
 	}
+}
 
-	_, err = c.protoClient.PublishEvent(c.withAuthToken(ctx), envelop)
+// PublishEventfromCustomContent serializes an struct and publishes its contents as data (JSON) onto topic in specific pubsub component.
+// DEPRECATED: This method is deprecated and will be removed in a future version of the SDK. Please use `PublishEvent` instead.
+func (c *GRPCClient) PublishEventfromCustomContent(ctx context.Context, pubsubName, topicName string, data interface{}) error {
+	log.Println("DEPRECATED: client.PublishEventfromCustomContent is deprecated and will be removed in a future version of the SDK. Please use `PublishEvent` instead.")
 
+	// Perform the JSON marshaling here just in case someone passed a []byte or string as data
+	enc, err := json.Marshal(data)
 	if err != nil {
-		return errors.Wrapf(err, "error publishing event unto %s topic", topicName)
+		return errors.WithMessage(err, "error serializing input struct")
```

Comment:
LGTM overall, just need add json marshal releated tests.

Reply:
@ItalyPaleAle please fix the above and we can merge :)

Reply:
Sorry I missed this notification.

If your ask was to add a test for this line, I've added it now.

However... This line was in previously-written code ([L46](https://github.com/dapr/go-sdk/pull/179/files#diff-7101101540524cbdbaa7b77971417897e57cb83c73d452bd214881bc22dae88cL46)) which existed before this PR It's also in a now-deprecated method. But ok :)